### ISSUE TITLE
gazelle cmd should be statically linkable to ease development

### ISSUE THREAD

**Issue by dragonsinth** [bazelbuild/bazel-gazelle#1291]
For people who are extending gazelle with other languages, IDE development is a chore.  There's no easy way to produce a working `gazelle` binary with a desired set of languages configured.

Proposal: 

1) Move almost all of `//cmd/gazelle` into a sub-package, leaving only `main()` and `languages`.
2) Have the sub-package expose a single `Run()` entry point.

Reasoning:

Developers extending gazelle could then trivially create their own custom gazelle binary, depending on the gazelle packages, with customized languages and minimal fuss.  Having a custom gazelle binary that doesn't require bazel-driven code generation means developers can easily build and debug in an IDE.

**Comment by sluongng** [comment]
I have yet to understand the motivation here. What is the pain point that you are trying to solve?

> Having a custom gazelle binary that doesn't require bazel-driven code generation means developers can easily build and debug in an IDE.

It's not very clear to me what use case do you have in mind that is not possible with current setup.

> Having a custom gazelle binary that doesn't require bazel-driven code generation 

Keep in mind that this repository, similar with `rules_go`, is built by and made for Bazel.
This means that we would not be able to guarentee backward compatible for non-Bazel use case.
Although if it's possible and easy to do, then I am all for it 👍 .
`rules_go` itself has many packages that does not work with `go build` and/or `go mod` by default.

**Comment by dragonsinth** [comment]
> I have yet to understand the motivation here. What is the pain point that you are trying to solve?

I want to be able to easily debug and iterate on plugin/language development from an IDE.

> > Having a custom gazelle binary that doesn't require bazel-driven code generation means developers can easily build and debug in an IDE.
> 
> It's not very clear to me what use case do you have in mind that is not possible with current setup.

I want to be able to write a simple gazelle binary like this:

```go
gazelle_impl.Run(languages, wd, os.Args[1:])
```

Where `languages` includes the custom language I'm working on and lives in my own repo, without forking the entirety of `cmd/gazelle`, which is a big package.

The normal way of building gazelle with custom languages goes entirely through bazel and is not IDE friendly.

> Keep in mind that this repository, similar with `rules_go`, is built by and made for Bazel. This means that we would not be able to guarentee backward compatible for non-Bazel use case. Although if it's possible and easy to do, then I am all for it 👍 . `rules_go` itself has many packages that does not work with `go build` and/or `go mod` by default.

Gazelle is completely compatible with the normal Go build.  If that were not true, you couldn't
`go get github.com/bazelbuild/bazel-gazelle` and have it work correctly.

**Comment by dragonsinth** [comment]
@sluongng yes, that's exactly it.  Thank you for helping me describe this better!  I changed the title of the issue, is that better?

**Comment by dragonsinth** [comment]
#1292 is the proof of concept I built, but can definitely iterate on it.  I could go and update it to use `bazel-gazelle/runner` if you think that's a good place to put it.

**Comment by albertocavalcante** [comment]
@dragonsinth, I am also interested in this! Any chance we could land https://github.com/bazel-contrib/bazel-gazelle/pull/1292, or does it require rework since it's been a while? Thanks! cc: @fmeum

**Comment by fmeum** [comment]
@jayconrod is working on a v2 and we probably wouldn't want to expose more internals before that's done.

**Comment by dragonsinth** [comment]
Yeah we're still using a fork. :/

**Comment by jayconrod** [comment]
v2 basically will do this. I haven't pushed PRs yet, but all the fix / update code moves into [`v2/cmd/gazelle/update`](https://github.com/jayconrod/bazel-gazelle/tree/v2/v2/cmd/gazelle/update) so that `cmd/gazelle` and `v2/cmd/gazelle` can still be separate binaries that share most of their logic. We may not guarantee backward compatibility for that package though, so it may very much be "use at your own risk".

I don't really get the motivation for this issue though. Why not build a binary with `gazelle_binary`, then check it in or install it?

**Comment by dragonsinth** [comment]
> I don't really get the motivation for this issue though. Why not build a binary with `gazelle_binary`, then check it in or install it?

Personally, I will run gazelle (with my stuff compiled in) from my IDE and debug/step through code.  So supporting non-bazel build of gazelle is really nice.

**Comment by jayconrod** [comment]
You can do that with a binary compiled with `gazelle_binary` though, right? The configuration is a little different but as long as you can `dlv exec` it, it should work.
